File: 0046-permutations/0046-permutations.cpp
class Solution {
    public:
    void permute(vector<int>& nums,vector<vector<int>>&ans,vector<int>&temp,vector<int>&vis){
        if(nums.size()==temp.size()){
            ans.push_back(temp);
            return;
        }
        for(int i=0;i<nums.size();i++){
            if(vis[i]==0){
                vis[i]=1;
                temp.push_back(nums[i]);
                permute(nums,ans,temp,vis);
                
                temp.pop_back();
                vis[i]=0;
            }
        }
    }

    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>>ans;
        vector<int>temp;
        vector<int>vis(nums.size(),0);
       
        permute(nums,ans,temp,vis);
        return ans;
    }
};

File: 0046-permutations/0046-permutations.py
class Solution:
    def permute(self, nums: List[int]) -> List[List[int]]:
        res=[]
        def sol(idx,subset):
            if idx==len(nums):
                res.append(nums[:])
                return
            for i in range(idx,len(nums)):
                nums[idx],nums[i]=nums[i],nums[idx]
                sol(idx+1,nums)
                nums[idx],nums[i]=nums[i],nums[idx]
        sol(0,nums)
        return res